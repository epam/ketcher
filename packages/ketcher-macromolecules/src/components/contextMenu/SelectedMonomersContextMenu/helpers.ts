import {
  AmbiguousMonomer,
  BaseMonomer,
  getRnaBaseFromSugar,
  isRnaBaseOrAmbiguousRnaBase,
  KetAmbiguousMonomerTemplateSubType,
  RNABase,
  Sugar,
} from 'ketcher-core';

const getMonomersCode = (monomers: BaseMonomer[]) => {
  return monomers
    .map((monomer) => monomer.monomerItem.props.MonomerNaturalAnalogCode)
    .sort()
    .join('');
};

export const isSenseBase = (monomer: BaseMonomer | AmbiguousMonomer) => {
  const { monomerItem } = monomer;
  const isNaturalAnalogue =
    monomerItem.props.MonomerNaturalAnalogCode === 'A' ||
    monomerItem.props.MonomerNaturalAnalogCode === 'C' ||
    monomerItem.props.MonomerNaturalAnalogCode === 'G' ||
    monomerItem.props.MonomerNaturalAnalogCode === 'T' ||
    monomerItem.props.MonomerNaturalAnalogCode === 'U';
  if (isNaturalAnalogue) {
    return true;
  }
  if (!monomer.monomerItem.isAmbiguous) {
    return false;
  }

  if (
    (monomer as AmbiguousMonomer).subtype ===
    KetAmbiguousMonomerTemplateSubType.MIXTURE
  ) {
    return false;
  }

  const N1 = 'ACGT';
  const N2 = 'ACGU';
  const B1 = 'CGT';
  const B2 = 'CGU';
  const D1 = 'AGT';
  const D2 = 'AGU';
  const H1 = 'ACT';
  const H2 = 'ACU';
  const K1 = 'GT';
  const K2 = 'GU';
  const W1 = 'AT';
  const W2 = 'AU';
  const Y1 = 'CT';
  const Y2 = 'CU';
  const M = 'AC';
  const R = 'AG';
  const S = 'CG';
  const V = 'ACG';
  const ambigues = [
    N1,
    N2,
    B1,
    B2,
    D1,
    D2,
    H1,
    H2,
    K1,
    K2,
    W1,
    W2,
    Y1,
    Y2,
    M,
    R,
    S,
    V,
  ];
  const code = getMonomersCode((monomer as AmbiguousMonomer).monomers);
  return ambigues.some((v) => v === code);
};

export const isAntisenseCreationDisabled = (
  selectedMonomers: BaseMonomer[],
) => {
  return selectedMonomers?.some((selectedMonomer: BaseMonomer) => {
    const rnaBaseForSugar =
      selectedMonomer instanceof Sugar && getRnaBaseFromSugar(selectedMonomer);

    return (
      (selectedMonomer instanceof RNABase &&
        (selectedMonomer.hydrogenBonds.length > 0 ||
          selectedMonomer.covalentBonds.length > 1)) ||
      (isRnaBaseOrAmbiguousRnaBase(selectedMonomer) &&
        !isSenseBase(selectedMonomer)) ||
      (rnaBaseForSugar &&
        (rnaBaseForSugar.hydrogenBonds.length > 0 ||
          !isSenseBase(rnaBaseForSugar)))
    );
  });
};
