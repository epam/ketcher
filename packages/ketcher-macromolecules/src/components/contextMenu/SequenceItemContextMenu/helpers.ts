import { flatten, get } from 'lodash';
import {
  Nucleotide,
  LabeledNucleotideWithPositionInSequence,
  NodeSelection,
  NodesSelection,
} from 'ketcher-core';

const generateLabeledNucleotides = (
  selectionsFlatten: NodeSelection[],
): LabeledNucleotideWithPositionInSequence[] => {
  const labeledNucleotides: LabeledNucleotideWithPositionInSequence[] = [];

  for (const selection of selectionsFlatten) {
    const node = selection.node;
    if (!(node instanceof Nucleotide)) {
      continue;
    }

    const { nodeIndexOverall } = selection;
    labeledNucleotides.push({
      baseLabel: node.rnaBase.label,
      sugarLabel: node.sugar.label,
      phosphateLabel: node.phosphate.label,
      nodeIndexOverall,
    });
  }

  return labeledNucleotides;
};

// Generate menu title if selected one nucleotide
const generateNucleotideTitle = (nucleotide: Nucleotide) => {
  let tempTitle = '';

  for (const property of ['sugar', 'rnaBase', 'phosphate']) {
    const propertyVal = nucleotide[property];
    const label = get(propertyVal, 'monomerItem.label', '');

    if (property === 'rnaBase') {
      tempTitle += `(${label})`;
    } else {
      tempTitle += label;
    }
  }

  return tempTitle;
};

export const generateSequenceContextMenuProps = (
  selections?: NodesSelection,
) => {
  if (!selections?.length) return;

  const selectionsFlatten: NodeSelection[] = flatten(selections);
  let title: string;
  let isSelectedAtLeastOneNucleotide = false;
  let isSelectedOnlyNucleotides = true;
  const selectionsCount = selectionsFlatten.length;
  let isSequenceFirstsOnlyNucleotidesSelected = true;

  for (let i = 0; i < selectionsCount; i++) {
    const node = selectionsFlatten[i].node;
    if (node instanceof Nucleotide) {
      const nodeAttachmentR1 = node.sugar.attachmentPointsToBonds.R1;
      if (nodeAttachmentR1 !== null) {
        isSequenceFirstsOnlyNucleotidesSelected = false;
      }
      isSelectedAtLeastOneNucleotide = true;
    } else {
      isSequenceFirstsOnlyNucleotidesSelected = false;
      isSelectedOnlyNucleotides = false;
    }
  }

  // Set title based on selected elements
  if (
    selectionsCount === 1 &&
    selectionsFlatten[0].node instanceof Nucleotide
  ) {
    title = generateNucleotideTitle(selectionsFlatten[0].node);
  } else {
    const titleElementType = isSelectedOnlyNucleotides
      ? 'nucleotides'
      : 'elements';
    title = `${selectionsCount} ${titleElementType}`;
  }

  // Generate labeled elements for RNA Builder
  const selectedSequenceLabeledNucleotides =
    generateLabeledNucleotides(selectionsFlatten);

  return {
    title,
    isSelectedOnlyNucleotides,
    isSelectedAtLeastOneNucleotide,
    selectedSequenceLabeledNucleotides,
    isSequenceFirstsOnlyNucleotidesSelected,
  };
};
