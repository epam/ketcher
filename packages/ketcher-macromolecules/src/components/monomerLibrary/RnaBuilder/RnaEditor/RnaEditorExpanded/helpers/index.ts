export * from './sequenceEdit';
export * from './resetRnaBuilder';
