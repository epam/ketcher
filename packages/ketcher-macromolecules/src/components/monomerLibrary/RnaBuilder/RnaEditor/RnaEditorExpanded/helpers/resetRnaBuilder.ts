import { AnyAction, Dispatch } from 'redux';
import { CoreEditor } from 'ketcher-core';
import {
  setActivePresetMonomerGroup,
  setIsEditMode,
  setSequenceSelection,
} from 'state/rna-builder';

const resetRnaBuilderCommon = (dispatch: Dispatch<AnyAction>) => {
  dispatch(setActivePresetMonomerGroup(null));
  dispatch(setIsEditMode(false));
};

export const resetRnaBuilder = (dispatch: Dispatch<AnyAction>) => {
  resetRnaBuilderCommon(dispatch);
};

export const resetRnaBuilderAfterSequenceUpdate = (
  dispatch: Dispatch<AnyAction>,
  editor: CoreEditor,
) => {
  resetRnaBuilderCommon(dispatch);
  dispatch(setSequenceSelection([]));
  editor.events.turnOffSequenceEditInRNABuilderMode.dispatch();
};
