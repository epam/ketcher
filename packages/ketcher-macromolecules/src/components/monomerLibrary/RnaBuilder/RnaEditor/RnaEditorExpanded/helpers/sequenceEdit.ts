import { LabeledNucleotideWithPositionInSequence } from 'ketcher-core';

const getNucleotideMonomerGroupName = (nameSet: Set<string>): string => {
  return nameSet.size === 1 ? [...nameSet][0] : '[multiple]';
};

export const generateSequenceSelectionGroupNames = (
  labeledNucleotides?: LabeledNucleotideWithPositionInSequence[],
) => {
  if (!labeledNucleotides?.length) return;

  const namesSets = {
    sugarLabel: new Set<string>(),
    baseLabel: new Set<string>(),
    phosphateLabel: new Set<string>(),
  };

  for (let i = 0; i < labeledNucleotides.length; i++) {
    for (const item of ['sugarLabel', 'baseLabel', 'phosphateLabel'])
      namesSets[item].add(labeledNucleotides[i]?.[item]);
  }

  return {
    Sugars: getNucleotideMonomerGroupName(namesSets.sugarLabel),
    Bases: getNucleotideMonomerGroupName(namesSets.baseLabel),
    Phosphates: getNucleotideMonomerGroupName(namesSets.phosphateLabel),
  };
};

export const generateSequenceSelectionName = (
  labeledNucleotide: LabeledNucleotideWithPositionInSequence,
) => {
  return `${labeledNucleotide?.sugarLabel}(${labeledNucleotide.baseLabel})${labeledNucleotide?.phosphateLabel}`;
};
