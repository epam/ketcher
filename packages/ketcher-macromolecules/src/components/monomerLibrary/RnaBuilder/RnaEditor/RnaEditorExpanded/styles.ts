/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import styled from '@emotion/styled';
import { Button, Input } from 'ketcher-react';

export const RnaEditorExpandedContainer = styled.div({
  display: 'flex',
  flexDirection: 'column',
  padding: '10px',
});

export const GroupsContainer = styled.div({
  marginTop: '16px',
  width: '100%',
  display: 'flex',
  flexDirection: 'column',
  gap: '8px',
});

export const ButtonsContainer = styled.div({
  marginTop: '16px',
  display: 'flex',
  justifyContent: 'space-evenly',
  alignItems: 'center',
  gap: '8px',
});

export const StyledButton = styled(Button)<{ primary?: boolean }>((props) => {
  return {
    width: '100%',
    whiteSpace: 'nowrap',
    fontSize: props.theme.ketcher.font.size.regular,
    backgroundColor: props.primary
      ? props.theme.ketcher.color.button.primary.active
      : undefined,
    color:
      props.primary && !props.disabled
        ? props.theme.ketcher.color.button.text.primary
        : props.theme.ketcher.color.text.light,
    outline:
      props.primary && !props.disabled
        ? props.theme.ketcher.outline.selected.small
        : props.theme.ketcher.outline.grey.small,
  };
});

export const NameContainer = styled.div<{ selected?: boolean }>((props) => ({
  position: 'relative',
  borderRadius: props.theme.ketcher.border.radius.regular,
  backgroundColor: props.theme.ketcher.color.background.primary,
  boxShadow: props.theme.ketcher.shadow.regular,
  cursor: 'pointer',
  overflow: 'hidden',
  padding: '0 6px 6px 6px',
  display: 'flex',
  alignItems: 'flex-end',
  height: '48px',
  outline: props.selected
    ? props.theme.ketcher.outline.selected.medium
    : 'none',

  ':hover': {
    outline: props.selected ? undefined : props.theme.ketcher.outline.small,
  },
}));

export const NameLine = styled.span<{ selected?: boolean }>((props) => ({
  position: 'absolute',
  top: '0',
  left: '0',
  width: '100%',
  height: '8px',
  backgroundColor: props.selected
    ? props.theme.ketcher.outline.selected.color
    : props.theme.ketcher.outline.color,
}));

export const NameInput = styled(Input)({
  width: '100%',
});

export const PresetName = styled.div({
  overflow: 'hidden',
  textOverflow: 'ellipsis',
});
