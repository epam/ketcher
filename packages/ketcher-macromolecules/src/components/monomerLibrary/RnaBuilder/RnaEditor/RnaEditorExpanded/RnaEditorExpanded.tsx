/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import { Entities, ModeTypes } from 'ketcher-core';
import { MonomerGroups } from 'src/constants';
import { GroupBlock } from './GroupBlock';
import {
  ButtonsContainer,
  GroupsContainer,
  NameContainer,
  NameInput,
  NameLine,
  PresetName,
  RnaEditorExpandedContainer,
  StyledButton,
} from './styles';
import { IRnaEditorExpandedProps } from 'components/monomerLibrary/RnaBuilder/RnaEditor/RnaEditorExpanded/types';
import { useDispatch } from 'react-redux';
import {
  monomerGroupToPresetGroup,
  RnaBuilderPresetsItem,
  savePreset,
  selectActivePreset,
  selectSequenceSelection,
  selectSequenceSelectionName,
  selectCurrentMonomerGroup,
  selectActivePresetMonomerGroup,
  selectActiveRnaBuilderItem,
  selectIsPresetReadyToSave,
  selectAllPresets,
  setActivePreset,
  setActiveRnaBuilderItem,
  setSugarValidations,
  setBaseValidations,
  setPhosphateValidations,
  setIsEditMode,
  selectPresetFullName,
  setUniqueNameError,
  setSequenceSelection,
  setSequenceSelectionName,
  selectIsActivePresetNewAndEmpty,
} from 'state/rna-builder';
import { useAppSelector, useLayoutMode } from 'hooks';
import {
  scrollToSelectedMonomer,
  scrollToSelectedPreset,
} from 'components/monomerLibrary/RnaBuilder/RnaEditor/RnaEditor';
import { getMonomerUniqueKey } from 'state/library';
import {
  selectEditor,
  selectIsSequenceEditInRNABuilderMode,
} from 'state/common';
import { ChangeEvent, useEffect, useState } from 'react';
import {
  generateSequenceSelectionGroupNames,
  generateSequenceSelectionName,
  resetRnaBuilder,
  resetRnaBuilderAfterSequenceUpdate,
} from 'components/monomerLibrary/RnaBuilder/RnaEditor/RnaEditorExpanded/helpers';
import { openModal } from 'state/modal';
import { getCountOfNucleoelements } from 'helpers/countNucleoelents';
import { getValidations } from 'helpers/rnaValidations';

type SequenceSelectionGroupNames = {
  [MonomerGroups.SUGARS]: string;
  [MonomerGroups.BASES]: string;
  [MonomerGroups.PHOSPHATES]: string;
};

export const RnaEditorExpanded = ({
  isEditMode,
  onDuplicate,
}: IRnaEditorExpandedProps) => {
  const groupsData = [
    {
      groupName: MonomerGroups.SUGARS,
      iconName: 'sugar',
      testId: 'rna-builder-slot--sugar',
    },
    {
      groupName: MonomerGroups.BASES,
      iconName: 'base',
      testId: 'rna-builder-slot--base',
    },
    {
      groupName: MonomerGroups.PHOSPHATES,
      iconName: 'phosphate',
      testId: 'rna-builder-slot--phosphate',
    },
  ] as const;

  const dispatch = useDispatch();
  const activePreset = useAppSelector(selectActivePreset);
  const isActivePresetEmpty = useAppSelector(selectIsActivePresetNewAndEmpty);
  const activeMonomerGroup = useAppSelector(selectActiveRnaBuilderItem);
  const editor = useAppSelector(selectEditor);
  const presets = useAppSelector(selectAllPresets);
  const activePresetMonomerGroup = useAppSelector(
    selectActivePresetMonomerGroup,
  );
  const [newPreset, setNewPreset] = useState(activePreset);

  // For sequence edit in RNA Builder mode
  const sequenceSelection = useAppSelector(selectSequenceSelection);
  const sequenceSelectionName = useAppSelector(selectSequenceSelectionName);
  const isSequenceEditInRNABuilderMode = useAppSelector(
    selectIsSequenceEditInRNABuilderMode,
  );
  const [isSequenceSelectionUpdated, setIsSequenceSelectionUpdated] =
    useState<boolean>(false);
  const [sequenceSelectionGroupNames, setSequenceSelectionGroupNames] =
    useState<SequenceSelectionGroupNames | undefined>(
      generateSequenceSelectionGroupNames(sequenceSelection),
    );

  const isSequenceMode = useLayoutMode() === ModeTypes.sequence;

  const updatePresetMonomerGroup = () => {
    if (activePresetMonomerGroup) {
      const groupName =
        monomerGroupToPresetGroup[activePresetMonomerGroup.groupName];
      const currentPreset = {
        ...newPreset,
        [groupName]: activePresetMonomerGroup.groupItem,
      };
      setNewPreset(currentPreset);
      return currentPreset;
    }
    return newPreset;
  };

  useEffect(() => {
    setNewPreset(activePreset);
  }, [activePreset]);

  useEffect(() => {
    if (!sequenceSelection) return;
    // If modifying 1 Nucleotide or 1 Nucleoside or Nucleoside with Phosphate in sequence
    if (getCountOfNucleoelements(sequenceSelection) === 1) {
      dispatch(
        setSequenceSelectionName(
          generateSequenceSelectionName(sequenceSelection),
        ),
      );
    }
    setSequenceSelectionGroupNames(
      generateSequenceSelectionGroupNames(sequenceSelection),
    );
  }, [dispatch, sequenceSelection]);

  useEffect(() => {
    if (activeMonomerGroup !== RnaBuilderPresetsItem.Presets && isEditMode) {
      if (isSequenceEditInRNABuilderMode && activePresetMonomerGroup) {
        const monomerType =
          monomerGroupToPresetGroup[activePresetMonomerGroup.groupName];
        const field = `${monomerType}Label`;
        const updatedSequenceSelection = sequenceSelection.map((node) => {
          // Do not set 'phosphateLabel' for Nucleoside if it is connected and selected with Phosphate
          // Do not set 'sugarLabel', 'baseLabel' for Phosphate
          if (
            (node.isNucleosideConnectedAndSelectedWithPhosphate &&
              field === 'phosphateLabel') ||
            (node.type === Entities.Phosphate &&
              (field === 'sugarLabel' || field === 'baseLabel'))
          )
            return node;

          return {
            ...node,
            [field]: activePresetMonomerGroup.groupItem.label,
          };
        });
        setIsSequenceSelectionUpdated(true);
        dispatch(setSequenceSelection(updatedSequenceSelection));
      } else {
        const currentPreset = updatePresetMonomerGroup();
        let presetFullName = newPreset?.name;
        if (!currentPreset.editedName) {
          presetFullName = selectPresetFullName(currentPreset);
        }
        setNewPreset({ ...currentPreset, name: presetFullName });
      }
    }
  }, [activePresetMonomerGroup?.groupItem, isSequenceEditInRNABuilderMode]);

  const scrollToActiveItemInLibrary = (selectedGroup) => {
    if (selectedGroup === RnaBuilderPresetsItem.Presets) {
      scrollToSelectedPreset(newPreset?.name);
      if (newPreset) {
        editor.events.selectPreset.dispatch(newPreset);
      }
      return;
    }

    const activeMonomerInSelectedGroup =
      newPreset[monomerGroupToPresetGroup[selectedGroup]];

    if (!activeMonomerInSelectedGroup) return;
    scrollToSelectedMonomer(getMonomerUniqueKey(activeMonomerInSelectedGroup));
  };

  const selectGroup = (selectedGroup) => () => {
    const selectedRNAPartMonomer = selectCurrentMonomerGroup(
      newPreset,
      selectedGroup,
    );

    if (selectedRNAPartMonomer && !isSequenceMode) {
      editor.events.selectMonomer.dispatch(selectedRNAPartMonomer);
    }
    scrollToActiveItemInLibrary(selectedGroup);
    dispatch(setActiveRnaBuilderItem(selectedGroup));

    const { sugarValidations, phosphateValidations, baseValidations } =
      getValidations(newPreset);

    dispatch(setSugarValidations(sugarValidations));
    dispatch(setPhosphateValidations(phosphateValidations));
    dispatch(setBaseValidations(baseValidations));
  };

  const onChangeName = (event: ChangeEvent<HTMLInputElement>) => {
    if (isEditMode) {
      const newPresetName = event.target.value;
      setNewPreset({
        ...newPreset,
        name: newPresetName.trim(),
        editedName: true,
      });
    }
  };

  const onUpdateSequence = () => {
    if (getCountOfNucleoelements(sequenceSelection) > 1) {
      dispatch(openModal('updateSequenceInRNABuilder'));
    } else {
      editor.events.modifySequenceInRnaBuilder.dispatch(sequenceSelection);
      resetRnaBuilderAfterSequenceUpdate(dispatch, editor);
    }
  };

  const onSave = () => {
    if (!newPreset?.name) {
      return;
    }
    const presetWithSameName = presets.find(
      (preset) => preset.name === newPreset.name,
    );
    if (
      presetWithSameName &&
      activePreset.nameInList !== presetWithSameName.name
    ) {
      dispatch(setUniqueNameError(newPreset.name));
      return;
    }
    dispatch(savePreset(newPreset));
    dispatch(setActivePreset(newPreset));
    editor.events.selectPreset.dispatch(newPreset);
    setTimeout(() => {
      scrollToSelectedPreset(newPreset.name);
    }, 0);
    resetRnaBuilder(dispatch);
  };

  const onCancel = () => {
    if (isSequenceEditInRNABuilderMode) {
      resetRnaBuilderAfterSequenceUpdate(dispatch, editor);
    } else {
      setNewPreset(activePreset);
      resetRnaBuilder(dispatch);
    }
  };

  const turnOnEditMode = () => {
    dispatch(setIsEditMode(true));
  };

  const getMonomerName = (groupName: string) => {
    if (
      activePresetMonomerGroup &&
      activePresetMonomerGroup.groupName === groupName
    ) {
      return activePresetMonomerGroup.groupItem.label;
    }
    return (
      selectCurrentMonomerGroup(newPreset, groupName)?.label ||
      selectCurrentMonomerGroup(newPreset, groupName)?.props.MonomerName
    );
  };

  const getMonomersName = (groupName: string) => {
    if (!sequenceSelectionGroupNames) return '';

    return sequenceSelectionGroupNames[groupName];
  };

  let mainButton;

  if (isActivePresetEmpty && !isSequenceEditInRNABuilderMode) {
    mainButton = (
      <StyledButton
        disabled={!selectIsPresetReadyToSave(newPreset)}
        primary
        data-testid="add-to-presets-btn"
        onClick={onSave}
      >
        Add to Presets
      </StyledButton>
    );
  } else if (isEditMode) {
    mainButton = (
      <StyledButton
        primary
        disabled={
          isSequenceEditInRNABuilderMode ? !isSequenceSelectionUpdated : false
        }
        data-testid="save-btn"
        onClick={isSequenceEditInRNABuilderMode ? onUpdateSequence : onSave}
      >
        {isSequenceEditInRNABuilderMode ? 'Update' : 'Save'}
      </StyledButton>
    );
  } else {
    mainButton = (
      <StyledButton
        data-testid="edit-btn"
        onClick={turnOnEditMode}
        disabled={activePreset.default}
      >
        Edit
      </StyledButton>
    );
  }

  return (
    <RnaEditorExpandedContainer
      data-testid="rna-editor-expanded"
      className={
        isSequenceEditInRNABuilderMode
          ? 'rna-editor-expanded--sequence-edit-mode'
          : ''
      }
    >
      <NameContainer
        selected={activeMonomerGroup === RnaBuilderPresetsItem.Presets}
        onClick={() => selectGroup(RnaBuilderPresetsItem.Presets)}
      >
        {isEditMode ? (
          <NameInput
            value={
              isSequenceEditInRNABuilderMode
                ? sequenceSelectionName
                : newPreset?.name
            }
            placeholder="Name your structure"
            disabled={isSequenceEditInRNABuilderMode}
            onChange={onChangeName}
          />
        ) : (
          <PresetName>{newPreset?.name}</PresetName>
        )}
        <NameLine
          selected={activeMonomerGroup === RnaBuilderPresetsItem.Presets}
        />
      </NameContainer>
      <GroupsContainer>
        {groupsData.map(({ groupName, iconName, testId }) => {
          return (
            <GroupBlock
              key={groupName}
              selected={activeMonomerGroup === groupName}
              groupName={groupName}
              monomerName={
                isSequenceEditInRNABuilderMode
                  ? getMonomersName(groupName)
                  : getMonomerName(groupName)
              }
              iconName={iconName}
              testid={testId}
              onClick={selectGroup(groupName)}
            />
          );
        })}
      </GroupsContainer>
      <ButtonsContainer>
        {isEditMode ? (
          <StyledButton data-testid="cancel-btn" onClick={onCancel}>
            Cancel
          </StyledButton>
        ) : (
          <StyledButton
            data-testid="duplicate-btn"
            disabled={!selectIsPresetReadyToSave(newPreset)}
            onClick={() => onDuplicate(newPreset)}
          >
            Duplicate and Edit
          </StyledButton>
        )}
        {mainButton}
      </ButtonsContainer>
    </RnaEditorExpandedContainer>
  );
};
