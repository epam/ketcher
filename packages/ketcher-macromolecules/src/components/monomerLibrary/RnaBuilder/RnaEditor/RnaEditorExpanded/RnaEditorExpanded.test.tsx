import { fireEvent, render, screen } from '@testing-library/react';
import { RnaEditorExpanded } from 'components/monomerLibrary/RnaBuilder/RnaEditor/RnaEditorExpanded/RnaEditorExpanded';
import { EmptyFunction } from 'helpers';

describe('Test Rna Editor Expanded component', () => {
  it('should render correctly in edit mode', async () => {
    render(
      withThemeAndStoreProvider(
        <RnaEditorExpanded isEditMode onDuplicate={EmptyFunction} />,
        {
          rnaBuilder: {
            activePreset: {
              name: 'MyRna',
              sugar: {
                props: {
                  MonomerName: '',
                },
              },
              phosphate: {
                props: {
                  MonomerName: '',
                },
              },
              base: {
                props: {
                  MonomerName: '',
                },
              },
            },
          },
        },
      ),
    );

    const rnaEditorExpanded = screen.getByTestId('rna-editor-expanded');
    const cancelBtn = screen.getByTestId('cancel-btn');
    const addToPresetsBtn = screen.getByTestId('add-to-presets-btn');

    fireEvent.click(addToPresetsBtn);
    fireEvent.click(cancelBtn);

    expect(rnaEditorExpanded).toMatchSnapshot();
  });

  it('should render correctly in view mode', async () => {
    const onDuplicateHandler = jest.fn();

    render(
      withThemeAndStoreProvider(
        <RnaEditorExpanded
          isEditMode={false}
          onDuplicate={onDuplicateHandler}
        />,
        {
          rnaBuilder: {
            activePreset: {
              name: 'MyRna',
              sugar: {
                props: {
                  MonomerName: '',
                },
              },
              phosphate: {
                props: {
                  MonomerName: '',
                },
              },
              base: {
                props: {
                  MonomerName: '',
                },
              },
              presetInList: {},
            },
          },
        },
      ),
    );

    const rnaEditorExpanded = screen.getByTestId('rna-editor-expanded');
    const editBtn = screen.getByTestId('edit-btn');
    const duplicateBtn = screen.getByTestId('duplicate-btn');

    fireEvent.click(editBtn);
    fireEvent.click(duplicateBtn);

    expect(onDuplicateHandler).toBeCalled();
    expect(rnaEditorExpanded).toMatchSnapshot();
  });
});
