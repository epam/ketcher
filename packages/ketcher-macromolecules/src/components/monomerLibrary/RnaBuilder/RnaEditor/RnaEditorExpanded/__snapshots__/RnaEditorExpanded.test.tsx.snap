// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Rna Editor Expanded component should render correctly in edit mode 1`] = `
<div
  class="css-zbq0po"
  data-testid="rna-editor-expanded"
>
  <div
    class="css-19lk7et"
  >
    <input
      class="css-9fgya1"
      placeholder="Name your structure"
      value="MyRna"
    />
    <span
      class="css-1o94icx"
    />
  </div>
  <div
    class="css-s3o4ys"
  >
    <div
      class="css-1b10s7n"
      data-testid="rna-builder-slot--sugar"
    >
      <svg
        class="css-1418ssm"
        fill="none"
        height="20"
        stroke-width="2"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="18"
          rx="1"
          width="18"
          x="1"
          y="1"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Sugar
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
    <div
      class="css-1b10s7n"
      data-testid="rna-builder-slot--base"
    >
      <svg
        class="css-1s5dbvi"
        fill="none"
        height="25"
        stroke-width="2"
        transform="translate(-3)"
        viewBox="0 0 26 25"
        width="26"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="15.013"
          rx="1"
          transform="rotate(45 13 1.414)"
          width="15.013"
          x="13"
          y="1.414"
        />
        <rect
          height="15.013"
          rx="1"
          stroke="currentColor"
          transform="rotate(45 13 1.414)"
          width="15.013"
          x="13"
          y="1.414"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Base
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
    <div
      class="css-1b10s7n"
      data-testid="rna-builder-slot--phosphate"
    >
      <svg
        class="css-b18d6u"
        fill="none"
        height="24"
        stroke-width="2"
        transform="translate(-2)"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="18"
          rx="9"
          width="18"
          x="3"
          y="3"
        />
        <rect
          height="18"
          rx="9"
          stroke="currentColor"
          width="18"
          x="3"
          y="3"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Phosphate
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
  </div>
  <div
    class="css-1iug9fi"
  >
    <button
      class="css-1971dyf"
      data-testid="cancel-btn"
    >
      Cancel
    </button>
    <button
      class="css-4dyvry"
      data-testid="add-to-presets-btn"
    >
      Add to Presets
    </button>
  </div>
</div>
`;

exports[`Test Rna Editor Expanded component should render correctly in view mode 1`] = `
<div
  class="css-zbq0po"
  data-testid="rna-editor-expanded"
>
  <div
    class="css-19lk7et"
  >
    <div
      class="css-1o4wo1x"
    >
      MyRna
    </div>
    <span
      class="css-1o94icx"
    />
  </div>
  <div
    class="css-s3o4ys"
  >
    <div
      class="css-1u2se0o"
      data-testid="rna-builder-slot--sugar"
    >
      <svg
        class="css-1418ssm"
        fill="none"
        height="20"
        stroke-width="2"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="18"
          rx="1"
          width="18"
          x="1"
          y="1"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Sugar
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
    <div
      class="css-1u2se0o"
      data-testid="rna-builder-slot--base"
    >
      <svg
        class="css-1s5dbvi"
        fill="none"
        height="25"
        stroke-width="2"
        transform="translate(-3)"
        viewBox="0 0 26 25"
        width="26"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="15.013"
          rx="1"
          transform="rotate(45 13 1.414)"
          width="15.013"
          x="13"
          y="1.414"
        />
        <rect
          height="15.013"
          rx="1"
          stroke="currentColor"
          transform="rotate(45 13 1.414)"
          width="15.013"
          x="13"
          y="1.414"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Base
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
    <div
      class="css-1u2se0o"
      data-testid="rna-builder-slot--phosphate"
    >
      <svg
        class="css-b18d6u"
        fill="none"
        height="24"
        stroke-width="2"
        transform="translate(-2)"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="18"
          rx="9"
          width="18"
          x="3"
          y="3"
        />
        <rect
          height="18"
          rx="9"
          stroke="currentColor"
          width="18"
          x="3"
          y="3"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Phosphate
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
  </div>
  <div
    class="css-1iug9fi"
  >
    <button
      class="css-1971dyf"
      data-testid="duplicate-btn"
    >
      Duplicate and Edit
    </button>
    <button
      class="css-1971dyf"
      data-testid="edit-btn"
    >
      Edit
    </button>
  </div>
</div>
`;
